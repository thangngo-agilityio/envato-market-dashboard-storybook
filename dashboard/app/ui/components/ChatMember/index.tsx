import { memo } from 'react';
import {
  Box,
  Flex,
  Text,
  useColorModeValue,
  useBreakpointValue,
} from '@chakra-ui/react';

// Themes
import { colors } from '@/ui/themes/bases/colors';

// Utils
import { generatePlaceholder, getStatusColor } from '@/lib/utils';

// Constants
import { IMAGES } from '@/lib/constants';
import Image from 'next/image';

export type Props = {
  avatar?: string;
  name?: string;
  localeTime?: string;
  icon?: React.ReactNode;
  statusColor?: string;
  onClick?: () => void;
  lastMessages?: string;
};

const ChatMember = ({
  avatar = IMAGES.CHAT_USER_AVATAR.url,
  name,
  lastMessages,
  localeTime,
  icon,
  statusColor = '',
  onClick,
}: Props) => {
  const colorFill = useColorModeValue(
    colors.secondary[200],
    colors.secondary[600],
  );

  const isMobile = useBreakpointValue({ base: true, lg: false });

  return isMobile ? (
    <Box
      cursor="pointer"
      _hover={{ bg: colorFill }}
      onClick={onClick}
      borderRadius="lg"
    >
      <Flex justify="space-between" p={3.5}>
        <Flex
          gap={3}
          borderRadius="50%"
          border="1px solid"
          w="48px"
          h="48px"
          position="relative"
        >
          <Image
            src={avatar || IMAGES.CHAT_USER_AVATAR.url}
            alt={avatar}
            width={48}
            height={48}
            placeholder="blur"
            blurDataURL={generatePlaceholder(48, 48)}
            style={{
              borderRadius: '50%',
            }}
          />

          <Box
            w="15px"
            h="15px"
            bg={getStatusColor(statusColor)}
            top={8}
            left={9}
            position="absolute"
            borderRadius="50%"
            border="3px solid"
            borderColor="common.white"
          />
        </Flex>

        <Flex direction="column" alignItems="center">
          <Text>{localeTime}</Text>
          {icon}
        </Flex>
      </Flex>
    </Box>
  ) : (
    <Box
      cursor="pointer"
      _hover={{ bg: colorFill }}
      onClick={onClick}
      borderRadius="lg"
    >
      <Flex justify="space-between" p={3.5} alignItems="flex-start">
        <Flex gap={3}>
          <Flex
            borderRadius="50%"
            minW="47px"
            h="47px"
            position="relative"
            mr={4}
          >
            <Image
              src={avatar || IMAGES.CHAT_USER_AVATAR.url}
              alt={avatar}
              width={47}
              height={47}
              placeholder="blur"
              blurDataURL={generatePlaceholder(47, 47)}
              style={{
                borderRadius: '50%',
              }}
            />

            <Box
              w="15px"
              h="15px"
              bg={getStatusColor(statusColor)}
              top="32px"
              left="35px"
              position="absolute"
              borderRadius="50%"
              border="3px solid"
              borderColor="common.white"
            />
          </Flex>
          <Flex flexDirection="column" alignItems="flex-start">
            <Box mr={6}>
              <Text fontSize="18px" fontWeight="bold">
                {name}
              </Text>
            </Box>
            <Text
              color="primary.300"
              dangerouslySetInnerHTML={{ __html: lastMessages ?? '' }}
            />
          </Flex>
        </Flex>

        <Flex direction="column" alignItems="center">
          <Text>{localeTime}</Text>
          {icon}
        </Flex>
      </Flex>
    </Box>
  );
};

const ChatMemberMemorized = memo(ChatMember);

export default ChatMemberMemorized;
