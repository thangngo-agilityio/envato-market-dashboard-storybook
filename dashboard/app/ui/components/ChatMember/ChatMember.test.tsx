import userEvent from '@testing-library/user-event';
import ChatMember from '../ChatMember';

const CHAT_MEMBER_PROPS = {
  name: 'Join',
  lastMessages: 'Hello',
  localeTime: '7h00',
};

describe('ChatMember component', () => {
  test('Should render match with snapshot.', () => {
    const { container } = render(<ChatMember {...CHAT_MEMBER_PROPS} />);

    expect(container).toMatchSnapshot();
  });

  test('call onClick when click member', async () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <ChatMember {...CHAT_MEMBER_PROPS} onClick={onClickMock} />,
    );

    const member = getByText('7h00');

    await userEvent.click(member);

    expect(onClickMock).toHaveBeenCalled();
  });
});
