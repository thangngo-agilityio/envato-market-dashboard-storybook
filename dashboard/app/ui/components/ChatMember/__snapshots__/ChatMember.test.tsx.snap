// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMember component Should render match with snapshot. 1`] = `
<div>
  <div
    class="css-1wloztz"
  >
    <div
      class="css-1826kcq"
    >
      <div
        class="css-it2hoo"
      >
        <img
          alt="/images/avatar-user-chat.webp"
          data-nimg="1"
          decoding="async"
          height="48"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Favatar-user-chat.webp&w=96&q=75"
          srcset="/_next/image?url=%2Fimages%2Favatar-user-chat.webp&w=48&q=75 1x, /_next/image?url=%2Fimages%2Favatar-user-chat.webp&w=96&q=75 2x"
          style="color: transparent; border-radius: 50%; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
          width="48"
        />
        <div
          class="css-nva1zn"
        />
      </div>
      <div
        class="css-dvxtzn"
      >
        <p
          class="chakra-text css-16h4vrm"
        >
          7h00
        </p>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
